# INNOMATICS HACKATHON TASK 


In [122]:
import pandas as pd
import numpy as np
import os

In [123]:
os.getcwd()

'D:\\SUSHMITHA\\Dataframes_csvFiles'

In [124]:
os.chdir("D:\SUSHMITHA\Dataframes_csvFiles")

# importing dataset

In [125]:
df=pd.read_csv("innomatics_data.csv")

In [126]:
df

x1            x2  y
0    -119.366669  1.150000e+02  1
1    -101.108044  9.777716e+01  1
2    -130.278658  1.067677e+02  1
3    -114.703415  1.011955e+02  1
4    -119.366669  1.150000e+02  1
...          ...           ... ..
2222   98.714112  8.964312e+01  0
2223   96.633331  9.100000e+01  0
2224   85.673940  1.038393e+02  0
2225   78.958862  7.860537e+01  0
2226  109.624009  1.000000e+08  0

[2227 rows x 3 columns]

In [127]:
df.shape

(2227, 3)

In [158]:
df.dtypes

x1    float64
x2    float64
y       int64
dtype: object

In [162]:
df.columns

Index(['x1', 'x2', 'y'], dtype='object')

# data preprocessing

In [128]:
df.isnull().any()

x1     True
x2     True
y     False
dtype: bool

In [129]:
df.isnull().sum()

x1    5
x2    3
y     0
dtype: int64

In [130]:
df=df.dropna()

In [131]:
df.shape

(2219, 3)

# Splitting of dataset

In [134]:
from sklearn.model_selection import train_test_split

In [135]:
X=df.drop(['y'],axis="columns")
y=df.y

In [136]:
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=2)

In [137]:
X_train


x1          x2
970    77.800201 -120.570212
953    67.633331 -114.000000
1441  -75.261003 -115.791311
1503  104.731254   76.804707
1139    2.650081  -29.815202
...          ...         ...
1075   39.856016   43.462461
437   -94.197970   77.821329
678   115.633331 -107.000000
1105   34.633331   17.000000
1616  -34.894961  -65.546649

[1775 rows x 2 columns]

In [138]:
X_train.shape

(1775, 2)

In [163]:
X_test

x1          x2
405   -74.824550   55.058290
1628  -49.612421  -78.175072
734   -76.910371  107.860296
1260 -126.410817 -120.840845
1093   38.633331   26.000000
...          ...         ...
740   -69.909481   96.122069
1570   21.101879   28.375560
402   -75.366669   50.000000
556   -41.291071   54.451348
125   -39.738155   31.625228

[444 rows x 2 columns]

In [139]:
X_test.shape

(444, 2)

In [164]:
y_train

970     1
953     1
1441    0
1503    0
1139    0
       ..
1075    0
437     1
678     1
1105    0
1616    0
Name: y, Length: 1775, dtype: int64

In [140]:
y_train.shape

(1775,)

In [165]:
y_test

405     1
1628    0
734     1
1260    0
1093    0
       ..
740     1
1570    0
402     1
556     1
125     1
Name: y, Length: 444, dtype: int64

In [141]:
y_test.shape

(444,)

# Creating a model

# decision tree

In [142]:
from sklearn import tree

In [143]:
model1=tree.DecisionTreeClassifier()

In [144]:
model1.fit(X_train,y_train)

DecisionTreeClassifier()

In [145]:
model1.score(X_test,y_test)

0.9346846846846847

# Logistic Regression

In [146]:
from sklearn.linear_model import LogisticRegression

In [147]:
model2=LogisticRegression()

In [148]:
model2.fit(X_train,y_train)

LogisticRegression()

In [149]:
model2.score(X_test,y_test)

0.6013513513513513

# KKN

In [150]:
from sklearn.neighbors import KNeighborsClassifier

In [179]:
model3=KNeighborsClassifier(n_neighbors=2)

In [180]:
model3.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=2)

In [182]:
model3.score(X_test,y_test)

0.9436936936936937

# SVM with Linear Kernal

In [154]:
from sklearn import svm

In [183]:
model4=svm.SVC(kernel='linear')

In [184]:
model4.fit(X_train,y_train)

SVC(kernel='linear')

In [185]:
model4.score(X_test,y_test)

0.5765765765765766

# SVM with RBF kernal

In [169]:
from sklearn import svm

In [176]:
model5=svm.SVC(kernel='rbf')

In [177]:
model5.fit(X_train,y_train)

SVC()

In [178]:
model5.score(X_test,y_test)

0.5765765765765766

# evaluate the model

In [186]:
y_test

405     1
1628    0
734     1
1260    0
1093    0
       ..
740     1
1570    0
402     1
556     1
125     1
Name: y, Length: 444, dtype: int64

In [189]:
X_test

x1          x2
405   -74.824550   55.058290
1628  -49.612421  -78.175072
734   -76.910371  107.860296
1260 -126.410817 -120.840845
1093   38.633331   26.000000
...          ...         ...
740   -69.909481   96.122069
1570   21.101879   28.375560
402   -75.366669   50.000000
556   -41.291071   54.451348
125   -39.738155   31.625228

[444 rows x 2 columns]

# For decision tree model

In [190]:
y_pred1=model1.predict(X_test)

In [191]:
y_pred1

array([1, 0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1,

In [193]:
df_predict1=pd.DataFrame({'Actual':y_test,'Predicted':y_pred1})

In [195]:
df_predict1

Actual  Predicted
405        1          1
1628       0          0
734        1          1
1260       0          0
1093       0          0
...      ...        ...
740        1          1
1570       0          1
402        1          1
556        1          1
125        1          1

[444 rows x 2 columns]

In [198]:
df_predict1['Actual'].equals(df_predict1['Predicted'])

False

In [200]:
x=df_predict1['Actual']==df_predict1['Predicted']

In [201]:
x

405      True
1628     True
734      True
1260     True
1093     True
        ...  
740      True
1570    False
402      True
556      True
125      True
Length: 444, dtype: bool

In [212]:
~df_predict1['Predicted'].values.sum()

-192

# for logistic regression model

In [214]:
y_pred2=model2.predict(X_test)

In [215]:
y_pred2

array([1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0,
       0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 0,
       0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1,

In [217]:
df_predict2=pd.DataFrame({'Actual':y_test,'Predicted':y_pred2})

In [218]:
df_predict2

Actual  Predicted
405        1          1
1628       0          0
734        1          1
1260       0          1
1093       0          0
...      ...        ...
740        1          1
1570       0          0
402        1          1
556        1          0
125        1          0

[444 rows x 2 columns]

In [222]:
x=df_predict2['Actual'].equals(df_predict2['Predicted'])

In [223]:
x

False

In [225]:
~df_predict2['Predicted'].values.sum()

-118

# for KNN model

In [226]:
y_pred3=model3.predict(X_test)

In [227]:
y_pred3

array([1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1,
       0, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0,
       1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 0, 0, 1, 1, 0, 0,
       0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1,

In [229]:
df_predict3=pd.DataFrame({'Actual':y_test,'Predicted':y_pred3})

In [230]:
df_predict3

Actual  Predicted
405        1          1
1628       0          0
734        1          1
1260       0          0
1093       0          0
...      ...        ...
740        1          1
1570       0          0
402        1          1
556        1          1
125        1          1

[444 rows x 2 columns]

In [231]:
x=df_predict3['Actual'].equals(df_predict3['Predicted'])

In [232]:
x

False

In [236]:
~df_predict3['Predicted'].values.sum()

-176

# for SVM linear  Model

In [247]:
y_pred4=model4.predict(X_test)

In [248]:
y_pred4

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [249]:
df_predict4=pd.DataFrame({'Actual':y_test,'Predicted':y_pred4})

In [250]:
df_predict4

Actual  Predicted
405        1          0
1628       0          0
734        1          0
1260       0          0
1093       0          0
...      ...        ...
740        1          0
1570       0          0
402        1          0
556        1          0
125        1          0

[444 rows x 2 columns]

In [251]:
x=df_predict4['Actual'].equals(df_predict4['Predicted'])

In [252]:
x

False

In [253]:
~df_predict4['Predicted'].values.sum()

-1

# for SVM RDF model

In [254]:
y_pred5=model5.predict(X_test)

In [255]:
y_pred5

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,

In [256]:
df_predict5=pd.DataFrame({'Actual':y_test,'Predicted':y_pred5})

In [257]:
df_predict5

Actual  Predicted
405        1          0
1628       0          0
734        1          0
1260       0          0
1093       0          0
...      ...        ...
740        1          0
1570       0          0
402        1          0
556        1          0
125        1          0

[444 rows x 2 columns]

In [258]:
x=df_predict5['Actual'].equals(df_predict5['Predicted'])

In [259]:
x

False

In [260]:
~df_predict5['Predicted'].values.sum()

-1

# observations

According to me, I think decision tree model is better than all other models
compared to other algorithms decision tree requires less effort for data preparation during preprocessing
it doesnot require any scaling of data as well
A decision tree model i very intuitive and easy to explain 
here,the missing values in the data also do not affect the process of bulding tree .
As observed,compared to all other model decision model gave an accuracy of 93% 